File: session-bg/plugins/fx/waves.lua
local W, H = 0, 0
local t = 0.0
local layers = {}

local SURFACE = { "▁", "▂", "▃", "▄", "▅", "▆", "▇", "█" }
local FILLS = { "·", "~", "≈", "░" }
local FOAM = "░"

local function surface_y(layer, x)
  local base = H * (1.0 - layer.depth)
  return base
    - layer.amp
      * (math.sin(x * layer.freq + t * layer.speed + layer.phase)
        + 0.5 * math.sin(x * layer.freq * 2.3 - t * layer.speed * 0.7))
end

function init(ctx)
  W, H = ctx.w, ctx.h
  t = 0.0
  layers = {}
  local rng = sbg.rng(ctx.seed)
  local palette = {
    { sbg.hex(0x2ac3de) },
    { sbg.hex(0x7aa2f7) },
    { sbg.hex(0x3d59a1) },
    { sbg.hex(0x1f2f5a) },
  }
  for i = 0, 3 do
    local rgb = palette[i + 1]
    layers[i + 1] = {
      depth = 0.34 - i * 0.08,
      amp = rng:range(0.8, 1.8) + i * 0.3,
      freq = rng:range(0.08, 0.16),
      speed = rng:range(0.6, 1.4) * (i % 2 == 0 and 1.0 or -1.0),
      phase = rng:range(0, 6.28318),
      rgb = rgb,
      fill = FILLS[i + 1],
    }
  end
end

function step(dt, state)
  t = t + dt
end

function render(fx, state)
  local density = state.params.density or 1.0
  local fill_rows = math.max(1, math.floor(2.0 * density + 0.5))
  local burst = state.mod.burst or 0.0
  for x = 0, W - 1 do
    local covered = H
    for i, layer in ipairs(layers) do
      local sy = surface_y(layer, x)
      local row = math.floor(sy)
      if row >= 0 and row < H and row < covered then
        local frac = 1.0 - (sy - math.floor(sy))
        local idx = sbg.clamp(math.floor(frac * 7.0 + 0.5), 0, 7) + 1
        local shade = 1.0 - (i - 1) * 0.18
        local r, g, b = sbg.scale(layer.rgb[1], layer.rgb[2], layer.rgb[3], 0.55 * shade)
        local ch = SURFACE[idx]
        if i == 1 and burst > 0.15 and sbg.smoothstep(0.0, 1.0, burst) > 0.5 then
          ch = FOAM
          r, g, b = sbg.scale(1.0, 1.0, 1.0, 0.5 * burst)
        end
        fx:put(x, row, ch, r, g, b)
        local y = row + 1
        while y < covered and y < row + 1 + fill_rows do
          local fr, fg, fb = sbg.scale(layer.rgb[1], layer.rgb[2], layer.rgb[3], 0.35 * shade)
          fx:put(x, y, layer.fill, fr, fg, fb)
          y = y + 1
        end
        covered = row
      end
    end
  end
end

File: session-bg/plugins/src/effects/waves.rs
use super::Effect;
use crate::frame::{hex, scale, Glyph};
use crate::rng::Rng;

const SURFACE: [char; 8] = ['▁', '▂', '▃', '▄', '▅', '▆', '▇', '█'];

struct Layer {
    depth: f32,
    amp: f32,
    freq: f32,
    speed: f32,
    phase: f32,
    rgb: [f32; 3],
    fill: char,
}

pub struct Waves {
    width: u16,
    height: u16,
    density: f32,
    t: f32,
    layers: Vec<Layer>,
}

impl Waves {
    pub fn new(density: f32) -> Self {
        Self {
            width: 0,
            height: 0,
            density,
            t: 0.0,
            layers: Vec::new(),
        }
    }

    fn surface_y(&self, layer: &Layer, x: f32) -> f32 {
        let h = f32::from(self.height);
        let base = h * (1.0 - layer.depth);
        base
            - layer.amp
                * ((x * layer.freq + self.t * layer.speed + layer.phase).sin()
                    + 0.5 * ((x * layer.freq * 2.3 - self.t * layer.speed * 0.7).sin()))
    }
}

impl Effect for Waves {
    fn resize(&mut self, width: u16, height: u16, rng: &mut Rng) {
        self.width = width;
        self.height = height;
        if self.layers.is_empty() {
            let palette = [hex(0x2ac3de), hex(0x7aa2f7), hex(0x3d59a1), hex(0x1f2f5a)];
            let fills = ['·', '~', '≈', '░'];
            for (i, rgb) in palette.iter().enumerate() {
                self.layers.push(Layer {
                    depth: 0.34 - i as f32 * 0.08,
                    amp: rng.range(0.8, 1.8) + i as f32 * 0.3,
                    freq: rng.range(0.08, 0.16),
                    speed: rng.range(0.6, 1.4) * if i % 2 == 0 { 1.0 } else { -1.0 },
                    phase: rng.range(0.0, std::f32::consts::TAU),
                    rgb: *rgb,
                    fill: fills[i],
                });
            }
        }
    }

    fn set_density(&mut self, density: f32) {
        self.density = density.clamp(0.1, 3.0);
    }

    fn step(&mut self, dt: f32, _rng: &mut Rng) {
        self.t += dt;
    }

    fn render(&self, out: &mut Vec<Glyph>) {
        let h = i32::from(self.height);
        let fill_rows = (2.0 * self.density).round().max(1.0) as i32;
        for x in 0..self.width {
            let mut covered = h;
            for (i, layer) in self.layers.iter().enumerate() {
                let sy = self.surface_y(layer, f32::from(x));
                let row = sy.floor() as i32;
                if row < 0 || row >= h {
                    continue;
                }
                if row >= covered {
                    continue;
                }
                let frac = 1.0 - (sy - sy.floor());
                let idx = ((frac * 7.0).round() as usize).min(7);
                let shade = 1.0 - i as f32 * 0.18;
                out.push(Glyph {
                    x,
                    y: row as u16,
                    ch: SURFACE[idx],
                    rgb: scale(layer.rgb, 0.55 * shade),
                });
                let mut y = row + 1;
                while y < covered && y < row + 1 + fill_rows {
                    out.push(Glyph {
                        x,
                        y: y as u16,
                        ch: layer.fill,
                        rgb: scale(layer.rgb, 0.35 * shade),
                    });
                    y += 1;
                }
                covered = row;
            }
        }
    }
}
